### DISCUSSION TITLE
audio-audio recording

### DISCUSSION THREAD

**Discussion by ZifengLIU143** [jspsych/jsPsych#3144] [Q&A] [answered]
Hi all, 

I am planning to run an audio recording experiment. In this experiment, participants will firstly see a text prompt about what they will say. After a very short interval (150 ms), they will hear a bib noise from the headset. Participants are asked to ignore the bib noise and speak out they have seen on the screen. For each trial, I want to start to calculate response times from the onset of the prompt to the end of the recording.

However, I did not find a plugin that supports audio as stimulus and allows audio recording. Do you have an idea to run this experiment in jspsych?

Thank you in advance! 

Zifeng

**Comment by ChristopheBossens** [comment, marked as answer]
Hi,

Would it be possible to combine an audio-button-response with an html-audio-response trial? The first one plays the bib noise and you can set it to proceed after the audio completes. I think jsPsych includes timestamps since the start of the experiment in its trails as well so you could use that to calculate response times?

Alternatively, could you use the html-audio-response and set the stimulus to HTML for playing an audio file yourself (not sure if this might have any timing implications):

```html
<audio controls autoplay>
  <source src="horse.ogg" type="audio/ogg">
</audio>
```

**Reply by ZifengLIU143** [reply]
Hi Christophe, 

Thank you for the response! I run a demo test by combining an audio-button-response with an html-audio-response trial. It works smoothly! But I am not sure about the accuracy of the response times:

1. The time gap calculated from two successive timestamps is not exactly the time gap I specify in the program, there is a tiny difference (~5 ms). Is this difference acceptable for RTs measurements? Do you know where does this difference come from?

2. How can I know when the recording starts? I notice that there is an "estimated stimulus onset" in the result. Does it mean the recording always starts a bit later than the display of stimulus? In addition, does using an external microphone affect the measurements of RTs, too? 

Apologize for so many questions ;)

Zifeng

**Reply by ZifengLIU143** [reply]
By the way, I found a custom plugin developed by [becky gilbert](https://github.com/becky-gilbert/jsPsych/blob/audio-response/plugins/jspsych-audio-audio-response.js). It may be useful, too. But I do not know how to use it. I downloaded the .js file in my folder, and add it with `<script>` tag in the script:
`<script src="jspsych-audio-audio-response.js"></script>`
But it did not work. Anyone can help?

**Reply by ChristopheBossens** [reply]
Hi Zifeng,

Unfortunately, these questions are a little bit too in-depth for me to answer accurately, but I can share my thoughts :). 

1. For the time gap that you mention, which timestamps are you referring to specifically? I think with online studies there will usually be some limits on your accuracy, and this should be considered in combination with the effect size you expect (e.g., if you expect ~200ms differences between conditions, then ~5ms is probably acceptable).

2. From the documentation I understand that the estimated stimulus onset refers to when the visual stimulus is displayed relative to the start of the audio recording, and the audio recording is always started before that (which means I would expect positive times here). If that is the case, the` time_elapsed` data attribute should correspond more or less to the start of the recording.

3. About the custom plugin: it seems to be developed in a previous version of jsPsych (6.3), could that be the reason why it does not work?

**Reply by ZifengLIU143** [reply]
Hi Christophe, 

Thank you for the answer! The custom plugin works now in a previous version. I still have a little concern about the timing.
For the time gap, I refer to the `time_elapsed`. For example, I specified a 500 ms fixation trial, a 200ms SOA, and an audio trial with a 525ms stimulus (=1225ms in total), but when I calculated the duration for these three trials based on the time elapsed, the result came out to be 1284ms, which is 59ms more. This difference mainly from the audio trial, it took a bit longer to finish the trial than the actual duration of the audio. I also find that this difference is quite arbitrary. I tried a couple of times and calculated the duration each time, the difference can be +5ms, +13ms, +45ms, +20ms, +28ms, etc. Do you have any idea about this? 

Best, 

Zifeng

**Comment by ZifengLIU143** [comment]
Hi @hesmall,
I try to use the custom audio-audio-response plugin in Jspsych 6.3, but I get an error: Uncaught TypeError: Failed to set the 'buffer' property on 'AudioBufferSourceNode': Failed to convert value to 'AudioBuffer'. Do you have any idea about this?
 
Thank you for your time! 

Zifeng
